### ISSUE TITLE
restored AWS VolumeID is missing availability zone

### ISSUE THREAD

**Issue by morganwalker** [heptio/velero#1249]
**What steps did you take and what happened:**

#### Take a backup
1. `kubectl config use-context cluster1`
2. Deployed the `nginx-app-with-pv.yaml` example to the `velero-restore-example` namespace. 
3. Show the PV that was created with `kubectl describe pv pvc-4cab9c94-3c2b-11e9-b93c-02d86058a9bc`
```
Name:            pvc-4cab9c94-3c2b-11e9-b93c-02d86058a9bc
Labels:          failure-domain.beta.kubernetes.io/region=us-east-1
                 failure-domain.beta.kubernetes.io/zone=us-east-1a
Annotations:     kubernetes.io/createdby: aws-ebs-dynamic-provisioner
                 pv.kubernetes.io/bound-by-controller: yes
                 pv.kubernetes.io/provisioned-by: kubernetes.io/aws-ebs
Finalizers:      [kubernetes.io/pv-protection]
StorageClass:    gp2
Status:          Bound
Claim:           velero-restore-example/nginx-logs
Reclaim Policy:  Delete
Access Modes:    RWO
Capacity:        1Gi
Node Affinity:   <none>
Message:
Source:
    Type:       AWSElasticBlockStore (a Persistent Disk resource in AWS)
    VolumeID:   aws://us-east-1a/vol-07cf5e71dab4e7bb6
    FSType:     ext4
    Partition:  0
    ReadOnly:   false
Events:         <none>
```

Note `VolumeID:   aws://us-east-1a/vol-07cf5e71dab4e7bb6`

3. `velero backup create nginx --include-namespaces 'velero-restore-example'`
completes successfully

#### Restore a backup
1. `kubectl config use-context cluster2`
2. `velero restore create --from-backup nginx`
3. `kubectl describe pv pvc-4cab9c94-3c2b-11e9-b93c-02d86058a9bc`
```
Name:            pvc-4cab9c94-3c2b-11e9-b93c-02d86058a9bc
Labels:          failure-domain.beta.kubernetes.io/region=us-east-1
                 failure-domain.beta.kubernetes.io/zone=us-east-1a
                 velero-restore=nginx-20190301105833
                 velero.io/backup-name=nginx
                 velero.io/restore-name=nginx-20190301105833
Annotations:     kubernetes.io/createdby: aws-ebs-dynamic-provisioner
                 pv.kubernetes.io/bound-by-controller: yes
                 pv.kubernetes.io/provisioned-by: kubernetes.io/aws-ebs
Finalizers:      [kubernetes.io/pv-protection]
StorageClass:    gp2
Status:          Bound
Claim:           velero-restore-example/nginx-logs
Reclaim Policy:  Delete
Access Modes:    RWO
Capacity:        1Gi
Node Affinity:   <none>
Message:
Source:
    Type:       AWSElasticBlockStore (a Persistent Disk resource in AWS)
    VolumeID:   vol-0959c4b96db0052b1
    FSType:     ext4
    Partition:  0
    ReadOnly:   false
Events:         <none>
```
Note `VolumeID:   vol-0959c4b96db0052b1`. The `StorageClass` was able to be restored because I'm using an image off of the latest master, gcr.io/heptio-images/velero@sha256:be5b270fb102e5b77307a37b6557f846baa2f8bd3dc9be0e8cb5c84475c90452.

**What did you expect to happen:**
The restore to create a volume with VolumeID `aws://<availability-zone>/<vol-xxxxx>`.  If I deploy `nginx-app-with-pv.yaml` directly to cluster2, I get the correct VolumeID.

**The output of the following commands will help us better understand what's going on**:

* `velero backup describe <backupname>` or `kubectl get backup/<backupname> -n velero -o yaml`
```
Name:         nginx
Namespace:    velero
Labels:       velero.io/storage-location=default
Annotations:  <none>

Phase:  Completed

Namespaces:
  Included:  velero-restore-example
  Excluded:  <none>

Resources:
  Included:        *
  Excluded:        <none>
  Cluster-scoped:  auto

Label selector:  <none>

Storage Location:  default

Snapshot PVs:  auto

TTL:  720h0m0s

Hooks:  <none>

Backup Format Version:  1

Started:    2019-03-01 10:57:32 -0500 EST
Completed:  2019-03-01 10:57:42 -0500 EST

Expiration:  2019-03-31 11:57:32 -0400 EDT

Validation errors:  <none>

Persistent Volumes:
  pvc-4cab9c94-3c2b-11e9-b93c-02d86058a9bc:
    Snapshot ID:        snap-06f9da638cfd291a6
    Type:               gp2
    Availability Zone:  us-east-1a
    IOPS:               <N/A>
```

* `velero restore describe <restorename>` or `kubectl get restore/<restorename> -n velero -o yaml`
```
Name:         nginx-20190301105833
Namespace:    velero
Labels:       <none>
Annotations:  <none>

Backup:  nginx

Namespaces:
  Included:  *
  Excluded:  <none>

Resources:
  Included:        *
  Excluded:        nodes, events, events.events.k8s.io, backups.ark.heptio.com, backups.velero.io, restores.ark.heptio.com, restores.velero.io
  Cluster-scoped:  auto

Namespace mappings:  <none>

Label selector:  <none>

Restore PVs:  auto

Phase:  Completed

Validation errors:  <none>

Warnings:  <none>
Errors:    <none>
```

**Anything else you would like to add:**
I'm trying to backup a namespace from a kops-provisioned 1.10.11 cluster and restore it to a CloudFormation-provisioned EKS 1.11.5 cluster. I'm using the same version of velero in both clusters but in cluster2 it's running with `--restore-only`. If the VolumeID isn't created correctly, our autoscaler (we're using Spotinst) fails to validate the volume and can't schedule the pod on a new node. We receive `PV Volume Id: vol-123456789 doesn't match volume pattern`.

**Environment:**

- Velero version (use `velero version`): v0.11.0
- Kubernetes version (use `kubectl version`): cluster1=1.10.11, cluster2=1.11.5
- Kubernetes installer & version: cluster1=kops 1.11.0, cluster2=EKS
- Cloud provider or hardware configuration: cluster1=AWS+kops, cluster2=AWS+EKS

**Comment by skriss** [comment]
@morganwalker so it sounds like everything's being restored OK, the volume comes up, etc. - but the only issue is that the PV's VolumeID is `vol-0959c4b96db0052b1` whereas you expect it to be `aws://us-east-1a/vol-0959c4b96db0052b1`? Am i reading that correctly?

**Comment by skriss** [comment]
Looking at the Kubernetes code, it looks like either form is valid, though the dynamic provisioner uses the `aws://<AZ>/<volume-id>` form.

**Comment by morganwalker** [comment]
The `nginx-app-with-pv.yaml` deployment and PVC were able to be successfully restored because the k8s node in cluster2 just happened to be in the same availability zone as the PVC that was restored, us-east-1a. If the k8s node in cluster2 was in us-east-1b, the nginx deployment and corresponding PVC would have been put into a pending state, which will trigger our autoscaler that a pod needs to be scheduled. The autoscaler performs validations and this particular one is failing with `PV Volume Id: vol-123456789 doesn't match volume pattern` because it's looking for `aws://<AZ>/<volume-id>`. Is there any reason why the dynamic provisioning form is being removed?

**Comment by skriss** [comment]
So two things - first, we don't natively support changing the AZ during a restore - the snapshot will always be restored to a volume in the same AZ as it came from. You might be able to write a plugin to do the remapping - not sure off the top of my head if it would work or not.

Second - I don't see any reason why we couldn't use the `aws://<AZ>/<volume-id>` form since from a Kubernetes perspective both are equally valid. Interested in contributing a patch?

**Comment by morganwalker** [comment]
> the snapshot will always be restored to a volume in the same AZ as it came from

Makes sense. When the snapshot is restored to a volume in us-east-1a and I don't currently have a node in us-east-1a, it should, and is, the job of the autoscaler to create a node in us-east-1a to accommodate the pending pod with a PVC in us-east-1a.

I'm willing to give a contribution a go! Any chance you could point me in the right direction of where I'd make the change?

**Comment by skriss** [comment]
Sure! It could be implemented in one of two places, but I'd probably modify this function: https://github.com/heptio/velero/blob/master/pkg/cloudprovider/aws/block_store.go#L275

The `volumeID` that's passed in is directly from AWS, so will look like `vol-xxxxx`.  You can grab the AZ of the volume from a label on the persistent volume -- similarly to how we do here: https://github.com/heptio/velero/blob/master/pkg/backup/item_backupper.go#L401.  Then you can use that to construct the `aws://<AZ>/<volume-id>`-formatted volume id.

Let me know if need any other input!

**Comment by skriss** [comment]
Closed by #1274
